src/watch/mod.rs

```diff
@@ -1,118 +1,87 @@
-    mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender},
-    oneshot,
-};
-use tonic::transport::Channel;
+
+use futures::stream::StreamExt;
+use smol::channel::{unbounded, Receiver, Sender};
+use smol::stream::Stream;
 
 pub use watch_impl::{WatchRequest, WatchResponse};
 
+use futures::prelude::*;
+
 use crate::lazy::{Lazy, Shutdown};
-use crate::proto::etcdserverpb;
-use crate::proto::etcdserverpb::watch_client::WatchClient;
-use crate::proto::mvccpb;
+use crate::protos::kv;
+use crate::protos::rpc;
+use crate::protos::rpc_grpc::WatchClient;
+use crate::KeyRange;
 use crate::KeyValue;
-use crate::Result;
-use crate::{Error, KeyRange};
+use crate::{Error, Result};
+use grpcio::WriteFlags;
 
 mod watch_impl;
 
 /// WatchTunnel is a reusable connection for `Watch` operation
 /// The underlying gRPC method is Bi-directional streaming
 struct WatchTunnel {
-    req_sender: UnboundedSender<WatchRequest>,
-    resp_receiver: Option<UnboundedReceiver<Result<WatchResponse>>>,
-    shutdown: Option<oneshot::Sender<()>>,
+    req_sender: Sender<WatchRequest>,
+    resp_receiver: Option<Receiver<Result<WatchResponse>>>,
+    shutdown: Option<Sender<()>>,
 }
 
 impl WatchTunnel {
-    fn new(mut client: WatchClient<Channel>) -> Self {
-        let (req_sender, mut req_receiver) = unbounded_channel::<WatchRequest>();
-        let (resp_sender, resp_receiver) = unbounded_channel::<Result<WatchResponse>>();
-
-        let request = tonic::Request::new(async_stream::stream! {
-            while let Some(req) = req_receiver.recv().await {
-                let pb: etcdserverpb::WatchRequest = req.into();
-                yield pb;
-            }
-        });
-
-        let (shutdown_tx, shutdown_rx) = oneshot::channel();
+    fn new(client: WatchClient) -> Self {
+        let (req_sender, req_receiver) = unbounded::<WatchRequest>();
+        let (resp_sender, resp_receiver) = unbounded::<Result<WatchResponse>>();
 
+        let (shutdown_tx, shutdown_rx) = unbounded();
+        let shutdown_reponse = shutdown_rx.clone();
         // monitor inbound watch response and transfer to the receiver
-        tokio::spawn(async move {
-            let mut shutdown_rx = shutdown_rx.fuse();
-            let mut inbound = futures::select! {
-                res = client.watch(request).fuse() => res.unwrap().into_inner(),
-                _ = shutdown_rx => { return; },
-            };
+        let (mut client_req_sender, mut resp) = client.watch().unwrap();
+        smol::spawn(async move {
+            let mut shutdown_rx = shutdown_rx.into_future().fuse();
+            while let Ok(req) = req_receiver.recv().await {
+                let watch_request: rpc::WatchRequest = req.into();
+
+                futures::select! {
+                    res = client_req_sender.send((watch_request, WriteFlags::default())).fuse() => res.unwrap(),
+                    _ = shutdown_rx => { return; }
+                };
+            }
+        })
+        .detach();
```

Comment:
if detach, you need to control the number of pending tasks

Reply:
how do you fix this issue?

Reply:
Add a new issue to track this.